astroquery/query.py

```diff
@@ -106,7 +107,35 @@ def from_cache(self, cache_location):
         return response
 
 
-@six.add_metaclass(abc.ABCMeta)
+class LoginABCMeta(abc.ABCMeta):
+    """
+    The goal of this metaclass is to copy the docstring and signature from
+    ._login methods, implemented in subclasses, to a .login method that is
+    visible by the users.
+
+    It also inherits from the ABCMeta metaclass as _login is an abstract
+    method.
+
+    """
+
+    def __new__(cls, name, bases, attrs):
+        newcls = super(LoginABCMeta, cls).__new__(cls, name, bases, attrs)
+
+        if '_login' in attrs and name not in ('BaseQuery', 'QueryWithLogin'):
+            # skip theses two classes, BaseQuery and QueryWithLogin, so
+            # below bases[0] should always be QueryWithLogin.
```

Comment:
Thinking about `bases[0]`, this could fail if there are multiple level of inheritance. That's probably not the case, but I need to find a better solution.